internal/controller/etcd/reconcile_delete.go

```diff
@@ -0,0 +1,117 @@
+	"github.com/gardener/etcd-druid/internal/utils"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// triggerDeletionFlow is the entry point for the deletion flow triggered for an etcd component which has a DeletionTimeStamp set on it.
+func (r *Reconciler) triggerDeletionFlow(ctx component.OperatorContext, logger logr.Logger, etcdObjectKey client.ObjectKey) ctrlutils.ReconcileStepResult {
+	deleteStepFns := []reconcileFn{
+		r.recordDeletionStartOperation,
+		r.deleteEtcdResources,
+		r.verifyNoResourcesAwaitCleanUp,
+		r.removeFinalizer,
+	}
+	for _, fn := range deleteStepFns {
+		if stepResult := fn(ctx, etcdObjectKey); ctrlutils.ShortCircuitReconcileFlow(stepResult) {
+			return r.recordIncompleteDeletionOperation(ctx, logger, etcdObjectKey, stepResult)
+		}
+	}
+	return ctrlutils.DoNotRequeue()
+}
+
+func (r *Reconciler) deleteEtcdResources(ctx component.OperatorContext, etcdObjKey client.ObjectKey) ctrlutils.ReconcileStepResult {
+	etcd := &druidv1alpha1.Etcd{}
+	if result := r.getLatestEtcd(ctx, etcdObjKey, etcd); ctrlutils.ShortCircuitReconcileFlow(result) {
+		return result
+	}
+	operators := r.operatorRegistry.AllOperators()
+	deleteTasks := make([]utils.OperatorTask, 0, len(operators))
+	for kind, operator := range operators {
+		// TODO: once we move to go 1.22 (https://go.dev/blog/loopvar-preview)
+		operator := operator
+		deleteTasks = append(deleteTasks, utils.OperatorTask{
+			Name: fmt.Sprintf("triggerDeletionFlow-%s-operator", kind),
+			Fn: func(ctx component.OperatorContext) error {
+				return operator.TriggerDelete(ctx, etcd)
+			},
+		})
+	}
+	ctx.Logger.Info("triggering triggerDeletionFlow operators for all resources")
+	if errs := utils.RunConcurrently(ctx, deleteTasks); len(errs) > 0 {
+		return ctrlutils.ReconcileWithError(errs...)
+	}
+	return ctrlutils.ContinueReconcile()
+}
+
+func (r *Reconciler) verifyNoResourcesAwaitCleanUp(ctx component.OperatorContext, etcdObjKey client.ObjectKey) ctrlutils.ReconcileStepResult {
+	etcd := &druidv1alpha1.Etcd{}
+	if result := r.getLatestEtcd(ctx, etcdObjKey, etcd); ctrlutils.ShortCircuitReconcileFlow(result) {
+		return result
+	}
+	operators := r.operatorRegistry.AllOperators()
+	resourceNamesAwaitingCleanup := make([]string, 0, len(operators))
+	for _, operator := range operators {
+		existingResourceNames, err := operator.GetExistingResourceNames(ctx, etcd)
+		if err != nil {
+			return ctrlutils.ReconcileWithError(err)
+		}
+		resourceNamesAwaitingCleanup = append(resourceNamesAwaitingCleanup, existingResourceNames...)
+	}
+	if len(resourceNamesAwaitingCleanup) > 0 {
+		ctx.Logger.Info("Cleanup of all resources has not yet been completed", "resourceNamesAwaitingCleanup", resourceNamesAwaitingCleanup)
+		return ctrlutils.ReconcileAfter(5*time.Second, "Cleanup of all resources has not yet been completed. Skipping removal of Finalizer")
```

Comment:
`has` is correct here, since it refers to `cleanup`, and not the `all resources` part of the sentence. So you would read it as `Cleanup (of all resources) has not yet been completed.`